src/cxx_support.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Support for `cxx` types.
+
+use std::pin::Pin;
+
+use cxx::memory::UniquePtrTarget;
+use cxx::UniquePtr;
+
+use crate::Emplace;
+use crate::TryNew;
+
+/// A type which has the ability to create heap storage space
+/// for itself in C++, without initializing that storage.
+///
+/// # Safety
+///
+/// Implementers must ensure that the pointer returned by
+/// `allocate_uninitialized_cpp_storage` is a valid, non-null,
+/// pointer to a new but uninitialized storage block, and that
+/// such blocks must be freeable using either of these routes:
+///
+/// * before they're initialized, using `free_uninitialized_cpp_storage`
+/// * after they're initialized, using the destructor of
+///   `std::unique_ptr<T, std::default_delete<T>>`
+pub unsafe trait MakeCppStorage
+where
+  Self: Sized,
+{
+  // Allocate heap space for this type in C++ and return a pointer
```

Comment:
Functions in `moveit` use indicative mood like Google C++ (i.e., "Allocates", not "Allocate").